File: server-practice/message-server/message-domain/domain-post/src/main/kotlin/dev/lkdcode/domain/aggregate/PostAggregate.kt
package dev.lkdcode.domain.aggregate

import dev.lkdcode.domain.entity.Post
import dev.lkdcode.domain.value.Author
import dev.lkdcode.domain.value.Content
import dev.lkdcode.domain.value.PostStatus
import dev.lkdcode.domain.value.Title
import java.time.Instant

class PostAggregate private constructor(
    private var post: Post
) {
    val getPost: Post get() = post

    fun update(
        newTitle: Title,
        newContent: Content,
        editor: Author,
    ): PostAggregate {
        this.post = post.update(newTitle, newContent, editor)

        return this
    }

    fun delete(
        deleter: Author
    ): PostAggregate {
        this.post = post.delete(deleter)

        return this
    }

    companion object {
        fun create(
            title: Title,
            content: Content,
            author: Author,
        ): PostAggregate = PostAggregate(
            Post(
                id = generatePostId(),
                title = title,
                content = content,
                author = author,
                status = PostStatus.DRAFT,
                viewCount = 0,
                createdAt = Instant.now(),
                updatedAt = Instant.now(),
            )
        )

        private fun generatePostId(): Post.PostId =
            Post.PostId(Instant.now().toEpochMilli())
    }
}

File: server-practice/py-server/py_domain/post/aggregate/post_aggregate.py
from datetime import datetime
from py_domain.post.entity.post import Post, PostId
from py_domain.post.value.title import Title
from py_domain.post.value.content import Content

class PostAggregate:
    def __init__(self, post: Post):
        self._post = post

    @property
    def post(self) -> Post:
        return self._post

    def update(self, title: Title, content: Content) -> 'PostAggregate':
        self._post.update(title, content)
        return self

    @classmethod
    def create(cls, title: str, content: str) -> 'PostAggregate':
        # Simple ID generation for example
        post_id = PostId(int(datetime.now().timestamp() * 1000))
        post = Post(
            id=post_id,
            title=Title(title),
            content=Content(content),
            created_at=datetime.now(),
            updated_at=datetime.now()
        )
        return cls(post)

